src/middleware/express.ts

```diff
@@ -0,0 +1,123 @@
+/*!
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as context from '@opencensus/propagation-stackdriver';
+import * as bunyan from 'bunyan';
+import {NextFunction, Request, Response} from 'express';
+
+import {LOGGING_TRACE_KEY, LoggingBunyan} from '../index';
+import * as types from '../types/core';
+
+export interface LoggingMiddlewareFunction {
+  (req: Request, res: Response, next: NextFunction): void;
+  logger: ReturnType<typeof bunyan.createLogger>;
+  mw: LoggingMiddlewareFunction;
+}
+
+export type LogObject = {
+  // tslint:disable-next-line:no-any bunyan interface.
+  [level in bunyan.LogLevelString]: (...args: any[]) => void;
+};
+
+export interface AnnotatedRequest extends Request {
+  log: LogObject;
+}
+
+function makeLogFunction(
+    level: bunyan.LogLevelString,
+    logger: ReturnType<typeof bunyan.createLogger>, trace: string|null) {
+  // tslint:disable-next-line:no-any bunyan interface.
+  return (...args: any[]) => {
+    let fields;
+    if (args[0] instanceof Error) {
+      // logger.info(err, ...);
+      fields = {err: args[0]};
+      args[0] = fields;
+    } else if (typeof args[0] !== 'object' || Array.isArray(args[0])) {
+      // logger.info(msg, ...);
+      fields = {};
+      args.unshift(fields);
+    } else {
+      // logger.info({...}, msg, ....);
+      fields = args[0];
+    }
+    if (trace) {
+      fields[LOGGING_TRACE_KEY] = trace;
+    }
+    return logger[level].apply(logger, args);
+  };
+}
+
+function makeLogObject(
+    logger: ReturnType<typeof bunyan.createLogger>,
+    trace: string|null): LogObject {
+  const log = {} as LogObject;
+  const levels: bunyan.LogLevelString[] =
+      ['trace', 'debug', 'info', 'warn', 'error', 'fatal'];
+  levels.forEach((level: bunyan.LogLevelString) => {
+    log[level] = makeLogFunction(level, logger, trace);
+  });
+  return log;
+}
+
+export interface ExpressMiddlewareOptions extends types.Options {
+  level?: types.LogLevel;
+}
+
+/**
+ * Express middleware
+ */
+export async function middleware(options?: ExpressMiddlewareOptions) {
```

Comment:
Annotate return type here? The return types can then be used as types in the tests below.

Reply:
Done.